scubainit/src/entfiles.rs

```diff
@@ -0,0 +1,205 @@
+use std::fs::File;
+use std::io::{BufRead, BufReader, Write};
+use std::marker::PhantomData;
+
+use crate::util::short_write;
+
+pub trait Entry: Sized {
+    fn from_line(line: &str) -> Result<Self, ReadEntryError>;
+    fn to_line(&self) -> String;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// ReadEntryError
+
+#[derive(Debug)]
+pub enum ReadEntryError {
+    Io(std::io::Error),
+    ParseInt(std::num::ParseIntError),
+    Invalid,
+}
+
+impl std::fmt::Display for ReadEntryError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            // The wrapped error contains additional information and is available
+            // via the source() method.
+            ReadEntryError::Io(_) => write!(f, "error reading entry from file"),
+
+            ReadEntryError::ParseInt(..) => write!(f, "an integer field could not be parsed"),
+
+            ReadEntryError::Invalid => write!(f, "the entry line is invalid"),
+        }
+    }
+}
+
+impl std::error::Error for ReadEntryError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match *self {
+            ReadEntryError::Io(ref e) => Some(e),
+            ReadEntryError::ParseInt(ref e) => Some(e),
+            _ => None,
+        }
+    }
+}
```

Comment:
I'd recommend taking a look at the [thiserror](https://crates.io/crates/thiserror) crate. It's very handy for making custom error types without having to write most of the boilerplate stuff that comes along with it.

I wouldn't say that you need to use it, but it's good know that it exists and can do a lot of the annoying aspects of writing a new error type for you.

Reply:
Great suggestion! I'd heard of *thiserror* before, but I assumed it was similar to *anyhow*, which I didn't want to use in the public API (in case I release the passwd/group/shadow code as its own create). But now I see that it really just helps derive implementations like I have with less code/effort.

Interestingly, this had negligible effect on the `scubainit` binary size:

| Mode | Before *thiserror* | After *thiserror* | Change |
|----------|----------------------------|--------------------------|-------------|
| Debug | 10980600 | 10982424 | +1824 |
| Release | 579320 | 579320 | +0 |

Reply:
Thanks a lot @jammiess for the helpful and educational review! 🍻 on me next time!